File: cpp/initialize.cpp
/*
Initialize functions including
- initialize_agents
- initialize_envs
- initialize_nbr_dict
- initialize_base_probs

March 21, 2018
SKG

Compilation command:
 */


#include "initialize.hpp"

/*
Set up an array of initial agent states
INPUTS:
N - total number of agents
T - total number of time steps from 0, 1, ..., T-1
K - total number of compartments
agent_status - T x N array where entry tn is agent n's status at time t
init_state_counts - 1d array of length K of the number of agents in each state at time 0.  That is entry k is the number of agents in state k at time 0.  Should add to N.
OUTPUTS:
agent_status array
 */
void initialize_agents(int N, int T, int K,
		  int agent_status[][1000],
		  int init_state_counts[]){
  // Initialize agents at time 0
  // so the first m_0 are in state 0, the next
  // m_1 are in state 1, ..., and the final m_K are in state K
  // The rest of the time steps are set to zero
  // WILL throw an error if there are fewer than 2 states
  int lower_bd = 0;
  int upper_bd = init_state_counts[0];
  for (int kk=0; kk < K; kk++){
    for (int ii=lower_bd; ii < upper_bd; ii++){
      agent_status[ii][0] = kk;
    }
    if (kk < K-1){
      lower_bd = upper_bd;
      upper_bd = upper_bd + init_state_counts[kk+1];
    }
  }




}



/*
  Initialize environments for the activities.  This is a N x E array where entry
  n,e is the nth agent's e^th environment assignment.  All 0s constitute a NULL assignment.
  @arg N number of agents
  @arg E number of environment types (e.g. schools + workplaces -> E =2)
  @arg max_env is the max number of different  categories
  @arg init_env_counts is a 2D E x max_env  array where entry j, i means that environment j has init_env_counts[i,j] agents with assignment i.
  @return a modified env variable, a N x E array.
 */
void initialize_envs(int N, int E, int max_env,
		     int env[][100],
		     int init_env_counts[][100]){

  for(int ee=0; ee < E; ee++){
    int lower_bd = 0;
    int upper_bd = init_env_counts[ee][0];
    for (int kk=0; kk < max_env; kk++){
      //     printf("upper_bd %d\n", upper_bd);
      if( upper_bd > N){
	break;
      }
      for (int ii=lower_bd; ii < upper_bd; ii++){
	env[ii][ee] = kk;
      }
      lower_bd = upper_bd;
      upper_bd = upper_bd + init_env_counts[ee][kk+1];
    }

  }


}


/*
Return graph (adjacency list) of neigbhors
INPUTS:
N - number of total agents
E - number of total environments
env - N x E array where entry ne is agent n's value for env e.  0 is a NULL assignment
OUTPUTS:
Graph object from boost library where edges represent being neighbors
 */

Graph initialize_nbr_graph(int N, int E, int env[][100]){
  Graph g(N);
  // For every agent
  // - Find neighbor of agent
  // - Add them to graph
  for(int ii=0; ii < N; ii++){

    for(int jj=0; jj < N; jj++){ // Loop over possible neighbors
      for(int ee=0; ee < E; ee++){ // Loop over environments
	int current_env = env[ii][ee];
	// NULL assignments (env =0) are NOT neighbors)
	// NEIGHBORS should not include self
	if( current_env > 0 & ii !=jj &
	    env[jj][ee] == current_env){
	  // Add neighbors to graph and break since we don't like repeat neighbors
	  boost::add_edge(ii, jj, g);
	  break;
	}
      }
    }

  }
  return g;
}

/*
Convert the initial CM transition rates to probabilities
INPUTS:
T - max amount of time so T+1 steps
K - number of states
params - (beta, gamma, N) all doubles
cm_vals - (T+1) x (K+1) matrix of times and CM values for each time step
base_probs - 3d array of dimension TxKxK. entry ijk is the probability of an agent in compartment j moving to compartment k from time i to i+1.
OUTPUTS:
updated base_probs 3d array
 */
void initialize_base_probs(int T, int K, std::array<double,3> params,
			     vector<state_type> cm_vals,
			     double base_probs[][100][100]){
  // TODO: MAKE MORE GENERIC

  for(int tt=0; tt < T; tt++){
    //printf("tt %d I %.2f N %.2f\n", tt, sir_vals[tt][2], N);
    double beta = params[0];
    double gamma = params[1];
    double N = params[2];
    
    base_probs[tt][0][1] = beta *
      cm_vals[tt][1] / N; // S to I
    base_probs[tt][0][2] = 0.0; // S to R
    base_probs[tt][0][0] = 1.0 - base_probs[tt][0][1]; // S to S
    base_probs[tt][1][0] = 0.0; // I to S
    base_probs[tt][1][1] = 1.0 - gamma; // I to I
    base_probs[tt][1][2] = gamma; // I to R
    base_probs[tt][2][0] = 0.0; // R to S
    base_probs[tt][2][1] = 0.0; // R to I
    base_probs[tt][2][2] = 1.0; // R to R
    
      
  }
}

File: catalyst/R/initialize.R
## Initialize agents, environment assignments, list of neighbors, 


#' Initialize the agents according to parameters in agent_list
#'
#' @param init_CM_vals vector of size K where entry k is the size of state k at time 0. 
#' @param N number of total agents
#' @param K number of total disease states
#' @param T max time, integer value
#' @return  (T+1) x N matrix of agent status with the first row (t=0) filled in with statuses corresponding to the init CM vals.  The other states are zero
initialize_agents <- function(init_CM_vals, N, K, T){
    
    agent_status <- matrix(0, nrow = (T + 1), ncol = N)
    if(sum(init_CM_vals) != N) stop("The initial CM values do not sum to N")
    if(length(init_CM_vals) != K) stop("There must be K initial values")
    ## Fill in initial values
    agent_status[1, ] <- rep(1:K, times  = init_CM_vals)
    return(agent_status)
}



#' Initialize environments
#'
#' @param env_list list with init_env_table a contingency table of number of individuals in each each cross section.  Each dimension name of the table corresponds to a number 1 to E_j, the number of total specific elements in Environment j.
#' @param N total number of agents
#' @param E total number of types of environments (e.g. schools and workplace -> E=2)
#' @return a NxE matrix where entry ij is agent i's value assignment to environment j.  An assignment of 0 corresponds to a NULL assignment.  Agents with value 0 do not belong to the same environment
initialize_env <- function(env_list, N, E){
    ## Convert table into data frame
    df <- as.data.frame(env_list$init_env_table,
                        stringsAsFactors = FALSE)
    ## Turn data frame into singular observations
    env_df <- df[rep((1:nrow(df))[df$Freq > 0],
                         times = df$Freq[df$Freq > 0]), 1:(ncol(df)-1), drop = FALSE]
    env_status <- as.matrix(env_df, drop = FALSE)
    env_status <- apply(env_status, 2, function(x) as.numeric(x))
    return(env_status)
}


#' Initialize the pre-computed neighbor list, which acts as a dictionary
#'
#' @param env_status a NxE matrix where entry ij is agent i's value assignment to environment j.  An assignment of 0 corresponds to a NULL assignment.  Agents with value 0 do not belong to the same environment
#' @param N total number of agents
#' @param E total number of environments
#' @return list of length N where entry n is a vector of indices of neighbors, that is people who share an environment assignment of non-zero for the same category.  The list entry has a 0 element if it has no neighbors
initialize_neighbors <- function(env_status, N, E){
    ## Initialize list
    neighbor_list <- vector(mode = "list", length = N)
    ## Loop over agents
    for(ii in 1:(N-1)){
        ## Loop over possible neighbors
        neighbor_inds <- NULL
        for(jj in (ii+1):N){
            ## Loop over environments

            for(ee in 1:E){
                current_env <- env_status[ii, ee]
                ## NULL assignments (env =0) are NOT neighbors
                ## NEIGHBORS should not include self
                if( current_env > 0  &
                    env_status[jj, ee] == current_env){
                    neighbor_list[[ii]] <- c(neighbor_list[[ii]], jj)
                    neighbor_list[[jj]] <- c(neighbor_list[[jj]], ii)
                    break
                }
            }
        }
        if(is.null(neighbor_list[[ii]])){
            neighbor_list[[ii]] <- 0
        }
    }
    if(is.null(neighbor_list[[N]])){
        neighbor_list[[N]] <- 0 ## Must manually check last entry
    }
  
    return(neighbor_list)
}


#' Find the initial probabilities of transition assuming homogeneity
#'
#' @param disease_list  List with
#' "params" - a numeric list of disease parameters (e.g. .1, .03)
#' "params_names" - optional variable names of params (e.g. beta, gamma)
#' "T" - max time.  The steps we take are 0, 1, ..., T inclusive
#' "init_vals" - vector of initial values of the states.  Must be nonnegative.  Length of vector is K, the number of states, and sum of values is N the constant number of individuals
#' @param CM_fxn output of make_CM_fxn
#' @return T x K x K matrix where entry t, i, j is the probability of an agent in compartment i transitioning to state j from time t-1 to t (because we start at 0)
#' @details TODO: currently hardcoded for SIR. Need to make general
initialize_probs <- function(disease_list, CM_fxn = SIR_diff){
    T <- disease_list$T
    K <- length(disease_list$init)
    N <- sum(disease_list$init)
    params <- disease_list$params
    

    ## Initialize array of dim TxKxK
    probs_mat <- array(0, c(T, K, K))

    ## Get ode_results
    ode_results <- sum_CM(disease_list, CM_fxn = CM_fxn)
    

    ## TODO: unhardcode

    ## If not specified, probability from i to j is 0
    probs_mat[, 1, 2] <-  params[1] * ode_results[1:T, 3] / N # S to I
    probs_mat[, 1, 1] <- 1 - probs_mat[, 1, 2] # S to S
    probs_mat[, 2, 3] <- params[2] # I to R
    probs_mat[, 2, 2] <- 1 - params[2] # I to I
    probs_mat[, 3, 3] <- 1 # R to R
           
    return(probs_mat)


}


#' Turn the base probs into individualized probs
#'
#' @param current_base_probs KxK matrix where entry ij is prob of transfer from i to j conditioned on current time step
#' @param current_agent_status integer vector of length N consisting of current status of agent
#' @return agent_probs N x K matrix where entry ni is probability of moving to state i conditioned on current state for agent n
base_to_agent_probs <- function(current_base_probs,
                                current_agent_status){
    agent_probs <- current_base_probs[current_agent_status,]
    return(agent_probs)

}
